tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs

```diff
@@ -526,88 +540,151 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
         }
 
         [SkippableTheory]
-        [ParameterAttributeData]
-        public void CustomEndpointTest([Values(false, true)] bool async)
+        // aws
+        [InlineData("aws", null, null, null)]
+        [InlineData("aws", "kms.us-east-1.amazonaws.com", null, null)]
+        [InlineData("aws", "kms.us-east-1.amazonaws.com:443", null, null)]
+        [InlineData("aws", "kms.us-east-1.amazonaws.com:12345", "$ConnectionRefusedSocketException$", null)]
+        [InlineData("aws", "kms.us-east-2.amazonaws.com", "us-east-1", null)]
+        [InlineData("aws", "example.com", "parse error", null)]
+        // additional not spec tests
+        [InlineData("aws", "$test$", "Invalid endpoint, expected dot separator in host, but got: $test$", null)]
+        // azure
+        [InlineData("azure", "key-vault-csfle.vault.azure.net", null, "parse error")]
+        // gcp
+        [InlineData("gcp", "cloudkms.googleapis.com:443", null, "parse error")]
+        [InlineData("gcp", "example.com:443", "Invalid KMS response", null)]
+        public void CustomEndpointTest(
+            string kmsType,
+            string customEndpoint,
+            string expectedExceptionInfoForValidEncryption,
+            string expectedExceptionInfoForInvalidEncryption)
         {
             RequireServer.Check().Supports(Feature.ClientSideEncryption);
 
-            using (var client = ConfigureClient())
-            using (var clientEncryption = ConfigureClientEncryption(client.Wrapped as MongoClient))
-            {
-                var testCaseMasterKey = new BsonDocument
-                {
-                    { "region", "us-east-1" },
-                    { "key", "arn:aws:kms:us-east-1:579766882180:key/89fcc2c4-08b0-4bd9-9f25-e30687b580d0" }
-                };
-                TestCase(testCaseMasterKey);
-
-                testCaseMasterKey = new BsonDocument
+            Action<string, Dictionary<string, object>> setValidKmsEndpointFunc =
```

Comment:
Why are these not local functions?

Reply:
because I pass these functions as argument here https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/157/files#diff-0537777d32b586e768ac1e74c10677b06f8933c41fba72c3f11a6f74862925e7R595-R596 (line 595-596)

Reply:
I understand that you pass them as arguments below.

My question is why they aren't a local function? That wouldn't prevent passing them would it?

Reply:
It's not a big deal... it just seemed odd that you have some `Action` variables that looked like they could/should be local functions.

Reply:
I didn't know that local functions can be used in this way. Thanks for your suggestion. Done